### DISCUSSION TITLE
Recommendations on using runStore(Entity)Action for client/server syncing

### DISCUSSION THREAD

**Discussion by xenobytezero** [salesforce/akita#630] [Q&A]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[ ] Feature request
[x] Documentation issue or request
[ ] Support request
[ ] Other... Please describe:
</code></pre>

## Request

Not sure how to categorise this, so I marked it as "documentation". Could be just a recommendation.

The model I have set up in my application essentially involves every action taken on a store to be sent to the server for persistence, and then relayed to the other connected clients. This is to prevent large data transfers constantly and to hopefully minimise collision.

Discovering Akita had `runStore(Entity)Action` was one of the primary reasons for using it, and have been very happy with how it has worked so far.

My original implementation of this pattern was to extend the `Store` and `EntityStore` to add a new `doSyncedChange` function which essentially used `runStore(Entity)Action` on the local store, then packed the params into a message to send to the server. This is then fired out to all the clients who then use `runStore(Entity)Action` to apply it to their own versions of the store.

This works, however you lose the ability to use the standard `update()`/`upsert()`/`remove()`/etc API, and all changes need to be done by this `doSyncedChange` function.

After updating to Akita v3 I decided to try and update this process a little to do this more transparently. As best I can tell it looks like I'm still going to have to extend the stores and override the main API, as well as resolve the data twice, which makes the code much more fragile.

It feels like there should be somewhere in the code I can hook in to a store to get "events" of some kind when the standard API is called, so I can package and send to the server, especially when the `persistState()`, `stateHistory` and middleware features exist.

In terms of documentation, the docs are very good at detailing how to handle incoming messages and apply them to stores, but doesn't really give a lot of clue on how to send those messages out.

Some guidance would be much appreciated, or am I barking up a terrible idea tree?

**Comment by xenobytezero** [comment]
I've just noticed the `selectEntityAction$` in `EntityStore`, which is about 90% of what I need, but that is not passed the new state. I would have to go retrieve the entity from ID and package the whole entity to send, rather than just the `Partial<>`.

**Reply by NetanelBasal** [reply]
That's correct.

**Reply by xenobytezero** [reply]
Before going down a rabbit hole I can never escape, does the Plugin API have lower level functionality I could use to build something that will get closer to what I'm after?

**Reply by NetanelBasal** [reply]
Can you describe exactly the data that you need?

**Reply by xenobytezero** [reply]
Essentially I would like to get the `Partial<S>` passed to an `update()`/`upsert()` call, so that I can send that `Partial` to all the other clients so they can run the same action on their side.

With the `selectEntityAction$` Observer I only get the ids of the changed entities, meaning I would need to query for the whole entity (for the second time, since it has already been done internally as part of the update) and send the whole entity to the clients which might be huge even though I have maybe only changed one property.

**Reply by NetanelBasal** [reply]
I see what you mean, but what if you pass a callback function?

```ts
store.update(1, state => {
  ...state,
   prop: { ...state.prop, ...updated }
})
```

**Reply by xenobytezero** [reply]
I think I'm not being as clear as I want to be. Let me try an example.

Assuming state `MyState` looks like this
```typescript
interface MyState {
    prop1: 'asdf'
    prop2: 'ghjk'
    prop3: [
        { item: '...' },
        { item: '...' },
        // x100
    ]
}
```

... and I have the following store ...

```typescript
@StoreConfig({ name: 'MyStore' })
export class MyStore extends EntityStore<MyState> {
}
```

... the rough version of my syncronisation query, which is sent to the server then relayed to every client looks like this ...

```typescript
export class MyQuery extends QueryEntity<MyState> {
    constructor(store: EntityStore<MyState>) {
        super(store);
        this.selectEntityAction().subscribe(action => {
            SyncService.syncToClients({
                store: this.store.storeName,
                action: action.type,
                data: {
                    ids: action.ids,
                    entities: action.ids.map(i => {
                        const e = this.getEntity(i);
                        if (e === undefined) { throw new Error('Cannot find entity - ' + i) };
                        return e;
                    })
                }
            })
        }
    }
}
```

As you can see, my only option for `Update` actions is to send the entire entity. meaning if I ran the following

```typescript
this.store.update((state) => {
    return {prop1: 'xyz'}
})
```

... my sync code above would get the action, then have to retrieve the entire entity to send to the other clients. The entity above has a giant array, which would have to be sent even though I'm only changing a single string parameter.

Ideally what I would want is to react to the `Update` action in the same way, but receive the `{prop1: 'xyz'}` returned from the update function, and just send that.

**Reply by NetanelBasal** [reply]
I understand what you're saying. My question was what you expect to be the "updated" object when you use the callback function: 

```ts
store.update(1, state => {
  ...state,
   prop: { ...state.prop, ...updated }
})
```

**Reply by xenobytezero** [reply]
Apologies for the delay, not been able to work on my project in a while.

I'm slightly confused as to what your asking. 

1. I'm not sure what the `updated` your referring to is? In my example above, I'm just passing back the `Partial<S>`  for the property I'm changing, which afaik is valid?
```
this.store.update((state) => {
    return { prop1: 'xyz' }
})
```

2. I could theoretically call a function in every `update()` callback to do the syncing, but that's a lot of boilerplate in something that feels like it should be hookable in a subclassed store.

**Reply by NetanelBasal** [reply]
What I'm saying is that the update method supports a callback function. I understand that you're not using it in your case, but others may be. So what you would expect to get as the "updated" object when someone uses the callback function.
